# 100 Node Questions (REST API - Authentication & Authorization)

##  Q1. What is Error Handling? In how many ways you can do error handling in Node.js?

Error handling is the process of managing errors that occur during the execution of a program or system.

### 4 ways to implement error handling in Node.js:
* Try-Catch (Sync)
* Error-First Callbacks (Async)
* Promises (Async)
* Try-Catch with async-await (Async)

##  Q2. How to handle errors in synchronous operations using try-catch-finally?

**TRY** – A try block is a block of code inside which any error can occur.

**CATCH** – When any error occurs in TRY block then it is passed to catch block to handle it.

**FINALLY** – The finally block is used to execute a given set of statements, whether an exception occur or not.

```javascript
try {
    // Synchronous operation that may throw an error
    throw new Error("Synchronous error");
} catch (error) {
    // Handle the error
    console.error("Error caught:", error.message);
} finally {
    // Code that runs regardless of whether an error occurred or not
    console.log("Finally block executed");
}

// Output: Error caught: Synchronous error
// Output: Finally block executed
```

##  Q3. What is Error-First Callbacks? `V.IMP.`

Error-First Callbacks is a convention in Node.js for handling asynchronous operations.

They are called Error-First Callbacks because the first argument of a callback function is reserved for an error object.

```javascript
// Define the error-first callback function
const errorFirstCallback = (error, result) => {
    if (error) {
        console.error("Error:", error.message);
        return;
    }
    console.log("Result:", result);
};

// Call the asynchronous operation and pass the error-first callback function in argument
asyncOperation(errorFirstCallback);

// Function simulating an asynchronous operation
function asyncOperation(callback) {
    // Simulate an asynchronous operation
    setTimeout(() => {
        const error = new Error("Async operation error");
        // Pass error as the first argument of the callback
        callback(error, null);
    }, 0);
}

// Output: Error: Async operation error
```

##  Q4. How to handle errors using Promises?

`.catch()` method is used in Promises for error handling.

```javascript
const asyncOperationPromise = new Promise((resolve, reject) => {
    // Perform asynchronous operation
    if (operationSuccessful) {
        resolve(result);
    } else {
        reject(new Error("Operation failed"));
    }
});

asyncOperationPromise
    .then((result) => console.log(result))
    .catch((error) => console.error(error.message));
```

##  Q5. How to handle errors while using async-await?

`try-catch` block is used with async-await for handling errors.

```javascript
// Simulating an asynchronous operation
function asyncOperation() {
  return new Promise((resolve, reject) => {
    // Perform asynchronous operation
    const operationSuccessful = Math.random() < 0.5;
    if (operationSuccessful) {
      resolve("Async operation result");
    } else {
      reject(new Error("Operation failed"));
    }
  });
}

// Async function to handle the asynchronous operation
async function handleAsyncOperation() {
  try {
    const result = await asyncOperation();
    console.log("Result:", result);
  } catch (error) {
    console.error("Error:", error.message);
  }
}

//Call the async function
handleAsyncOperation();
```

##  Q6. How can you Debug Node.js applications?

Debugging techniques in Node.js
* console.log()
* debugger statement
* Node.js inspector
* Visual Studio Code debugger
* Chrome DevTools

<!---
Adarsh 
2nd August 2024
08:46 AM
(18:09)
--->